### PR TITLE
Add gitian build script

### PR DESCRIPTION
**Pull request by TheCharlatan**
This adds a gitian build script and gitian descriptor to achieve reproducible builds in monero. 
Currently the build script in gitian-build.py still clones bitcoin's repository used to exchange, compare and store signatures. These two repos should be created on monero, namely gitian.sigs and ~monero-detached-sigs~. Edit: since no apple or windows certs are required, monero-detached-sigs is not needed. 
It also just builds Linux binaries at the moment. Gitian descriptors for OSX and windows still need to be created. 
To test this out, create a remote copy of this branch on your fork, tag a version to it in the format 'vN.N.N' and prepare the host as depicted in the README included in this pull request. Then setup and run a detached signature build.
Please post reported checksums below, so we see if the current setup already is reproducible, or if more tweaking is needed. I will post checksums as soon as my build is done.

### CODE REVIEW

**Comment by TheCharlatan** [comment]
I am getting the following output with this branch tagged at version v0.0.32 :

- 019f0bf2dae7fb415d5d624803dca4d54c4ad852b161fefc8c363b795d8910ed  monero-aarch64-linux-gnu.tar.gz
- 3fecde5ddf66de997210328fdf9a208a1d1aa403a63ad8a75bab62cd39e84a05  monero-arm-linux-gnueabihf.tar.gz
- 4cac79dd8b2403b67299b153bab1908fa7f4f3864f51e74cdf9aa040ebe1be08  monero-i686-linux-gnu.tar.gz
- 57a32fc10ff71b2dd445bf40a97a80b695a85c523780d5de1f5f34024eea7a17  monero-x86_64-linux-gnu.tar.gz
- 203a7a1a0debf3b98f1a82a319123f6ca178d88194fd0d3e5cdbc2d7521386f7  monero-linux-0.18-res.yml

**Comment by TheCharlatan** [comment]
Made a second run on the same machine and the outputs remained the same. :D

File: contrib/gitian/gitian-build.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        programs += ['lxc', 'debootstrap']
+    subprocess.check_call(['sudo', 'apt-get', 'install', '-qq'] + programs)
+    if not os.path.isdir('gitian.sigs'):
+        subprocess.check_call(['git', 'clone', 'https://github.com/bitcoin-core/gitian.sigs.git'])
+    if not os.path.isdir('monero-detached-sigs'):
+        subprocess.check_call(['git', 'clone', 'https://github.com/bitcoin-core/bitcoin-detached-sigs.git'])
+    if not os.path.isdir('gitian-builder'):
+        subprocess.check_call(['git', 'clone', 'https://github.com/devrandom/gitian-builder.git'])
+    if not os.path.isdir('monero'):
+        subprocess.check_call(['git', 'clone', 'https://github.com/bitcoin-core/monero.git'])
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply contrib/gitian/gitian-build.py]: I added a checkout here to make sure the gitian-builder repo was a specific commit.

[the remark this answers is not in the document]
Comment (TheCharlatan) [reply contrib/gitian/gitian-build.py]: Corrected.

File: contrib/gitian/gitian-build.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            exit(1)
+    else:
+        programs += ['lxc', 'debootstrap']
+    subprocess.check_call(['sudo', 'apt-get', 'install', '-qq'] + programs)
+    if not os.path.isdir('gitian.sigs'):
+        subprocess.check_call(['git', 'clone', 'https://github.com/bitcoin-core/gitian.sigs.git'])
+    if not os.path.isdir('monero-detached-sigs'):
+        subprocess.check_call(['git', 'clone', 'https://github.com/bitcoin-core/bitcoin-detached-sigs.git'])
+    if not os.path.isdir('gitian-builder'):
+        subprocess.check_call(['git', 'clone', 'https://github.com/devrandom/gitian-builder.git'])
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply contrib/gitian/gitian-build.py]: As a quick fix the script now checks out to the last commit hash.

File: contrib/gitian/gitian-build.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+    os.makedirs('monero-binaries/' + args.version, exist_ok=True)
+    print('\nBuilding Dependencies\n')
+    os.chdir('gitian-builder')
+    os.makedirs('inputs', exist_ok=True)
+
+    subprocess.check_call(['wget', '-N', '-P', 'inputs', 'http://downloads.sourceforge.net/project/osslsigncode/osslsigncode/osslsigncode-1.7.1.tar.gz'])
+    subprocess.check_call(['wget', '-N', '-P', 'inputs', 'https://bitcoincore.org/cfields/osslsigncode-Backports-to-1.7.1.patch'])
+    subprocess.check_call(['make', '-C', '../monero/contrib/depends', 'download', 'SOURCES_PATH=' + os.getcwd() + '/cache/common'])
+
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply contrib/gitian/gitian-build.py]: Changed the urls to https and added checks for their sha256sum.

**Comment by TheCharlatan** [comment]
I tested the builds for the following target platforms (all the current targets that binaries are distributed for) on the v0.0.59 tag on my github remote now: 
x86_64-linux-gnu
aarch64-linux-gnu
i686-linux-gnu
arm-linux-gnueabihf
x86_64-apple-darwin11
x86_64-w64-mingw32
i686-w64-mingw32

They consistently produced the same checksums for the binaries.

File: contrib/gitian/gitian-build.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        subprocess.check_call(['git', 'add', args.version+'-linux/'+args.signer])
+        subprocess.check_call(['git', 'add', args.version+'-win-unsigned/'+args.signer])
+        subprocess.check_call(['git', 'add', args.version+'-osx-unsigned/'+args.signer])
+        subprocess.check_call(['git', 'commit', '-m', 'Add '+args.version+' unsigned sigs for '+args.signer])
+        os.chdir(workdir)
+
+def sign():
+    global args, workdir
+    os.chdir('gitian-builder')
+    os.chdir(workdir)
```
Comment (moneromooo-monero) [inline contrib/gitian/gitian-build.py]: workdir seems to be an abolute path, from below. So the first chdir is ignored ?

File: contrib/gitian/gitian-build.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    print('\nVerifying v'+args.version+' Linux\n')
+    subprocess.check_call(['bin/gverify', '-v', '-d', '../gitian.sigs/', '-r', args.version+'-linux', '../monero/contrib/gitian/gitian-linux.yml'])
+    print('\nVerifying v'+args.version+' Windows\n')
+    subprocess.check_call(['bin/gverify', '-v', '-d', '../gitian.sigs/', '-r', args.version+'-win-unsigned', '../monero/contrib/gitian/gitian-win.yml'])
+    print('\nVerifying v'+args.version+' MacOS\n')
+    subprocess.check_call(['bin/gverify', '-v', '-d', '../gitian.sigs/', '-r', args.version+'-osx-unsigned', '../monero/contrib/gitian/gitian-osx.yml'])
+    print('\nVerifying v'+args.version+' Signed Windows\n')
+    subprocess.check_call(['bin/gverify', '-v', '-d', '../gitian.sigs/', '-r', args.version+'-win-signed', '../monero/contrib/gitian/gitian-win-signer.yml'])
+    print('\nVerifying v'+args.version+' Signed MacOS\n')
+    subprocess.check_call(['bin/gverify', '-v', '-d', '../gitian.sigs/', '-r', args.version+'-osx-signed', '../monero/contrib/gitian/gitian-osx-signer.yml'])
```
Comment (moneromooo-monero) [inline contrib/gitian/gitian-build.py]: Is it intended to have two chunks for OSX and Windows, but one for Linux ?

File: contrib/gitian/gitian-build.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        exit(1)
+    if args.version == '':
+        print(script_name+': Missing version.')
+        print('Try '+script_name+' --help for more information')
+        exit(1)
+
+    # Add leading 'v' for tags
+    if args.commit and args.pull:
+        raise Exception('Cannot have both commit and pull')
+    args.commit = ('' if args.commit else 'v') + args.version
```
Comment (moneromooo-monero) [inline contrib/gitian/gitian-build.py]: Why are tags specified with the first character removed ? This seems complicated and arbitrary, but there might be a good reason...

File: contrib/gitian/README.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+echo "%sudo ALL=NOPASSWD: /usr/bin/lxc-start" > /etc/sudoers.d/gitian-lxc
+echo "%sudo ALL=NOPASSWD: /usr/bin/lxc-execute" >> /etc/sudoers.d/gitian-lxc
+# make /etc/rc.local script that sets up bridge between guest and host
+echo '#!/bin/sh -e' > /etc/rc.local
+echo 'brctl addbr br0' >> /etc/rc.local
+echo 'ip addr add 10.0.3.1/24 broadcast 10.0.3.255 dev br0' >> /etc/rc.local
+echo 'ip link set br0 up' >> /etc/rc.local
+echo 'firewall-cmd --zone=trusted --add-interface=br0' >> /etc/rc.local
+echo 'exit 0' >> /etc/rc.local
+chmod +x /etc/rc.local
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply contrib/gitian/README.md]: I'm not sure what you mean. Are suggesting to use cat instead of echo and io redirects?

**Comment by TheCharlatan** [comment]
No, additional work will come as another pul request, since it should not be directly related to the scripts.

**Pull request merged**